File: engines/clocks/pong_clock.py
from PIL import Image, ImageDraw, ImageFont
from datetime import datetime
import random
import math

class PongClock:
    def __init__(self, width, height):
        self.w = width
        self.h = height
        self.ball_x = width // 2
        self.ball_y = height // 2
        self.ball_dx = 2.0
        self.ball_dy = 1.5
        self.ball_size = 2
        
        self.pad_h = 12
        self.pad_w = 2
        self.p1_y = (height - self.pad_h) // 2
        self.p2_y = (height - self.pad_h) // 2
        
        self.last_minute = -1
        self.force_miss = False
        
    def reset_ball(self, left_served):
        self.ball_y = self.h // 2
        self.ball_dy = random.uniform(-1.5, 1.5)
        if left_served:
            self.ball_x = self.pad_w + 1
            self.ball_dx = 2.0
        else:
            self.ball_x = self.w - self.pad_w - 3
            self.ball_dx = -2.0

    def tick(self, img, time_str, font, color1, color2):
        draw = ImageDraw.Draw(img)
        
        # Parse time
        parts = time_str.split(':')
        if len(parts) >= 2:
            score_left = parts[0]
            score_right = parts[1]
        else:
            score_left = "00"
            score_right = "00"
            
        now_min = int(score_right) if score_right.isdigit() else 0
        if self.last_minute == -1:
            self.last_minute = now_min
        elif self.last_minute != now_min:
            self.force_miss = True
            self.last_minute = now_min
            
        # Draw dotted middle line
        for y in range(0, self.h, 4):
            draw.line([(self.w//2, y), (self.w//2, y+1)], fill=(100, 100, 100))
            
        # Draw scores
        try:
            bbox_l = draw.textbbox((0, 0), score_left, font=font)
            w_l = bbox_l[2] - bbox_l[0]
        except:
            w_l = 15
            
        draw.text(((self.w//2) - w_l - 8, 4), score_left, font=font, fill=color1)
        draw.text(((self.w//2) + 8, 4), score_right, font=font, fill=color2)
        
        # Physics
        self.ball_x += self.ball_dx
        self.ball_y += self.ball_dy
        
        # Top/Bottom Bounce
        if self.ball_y <= 0:
            self.ball_y = 0
            self.ball_dy *= -1
        elif self.ball_y >= self.h - self.ball_size:
            self.ball_y = self.h - self.ball_size
            self.ball_dy *= -1
            
        # P1 AI (Left) - Perfect tracking
        target_p1 = self.ball_y - (self.pad_h // 2)
        if self.ball_dx < 0:
            if self.p1_y < target_p1: self.p1_y += 1.5
            if self.p1_y > target_p1: self.p1_y -= 1.5
            
        # P2 AI (Right) - Perfect unless force_miss
        target_p2 = self.ball_y - (self.pad_h // 2)
        if self.ball_dx > 0:
            if self.force_miss:
                # Move away from ball
                if self.ball_y > self.h // 2: target_p2 = 0
                else: target_p2 = self.h - self.pad_h
                
            if self.p2_y < target_p2: self.p2_y += 1.5
            if self.p2_y > target_p2: self.p2_y -= 1.5
            
        # Clamp paddles
        self.p1_y = max(0, min(self.h - self.pad_h, self.p1_y))
        self.p2_y = max(0, min(self.h - self.pad_h, self.p2_y))
        
        # Paddle collisions
        if self.ball_dx < 0 and self.ball_x <= self.pad_w:
            if self.p1_y - self.ball_size <= self.ball_y <= self.p1_y + self.pad_h:
                self.ball_x = self.pad_w
                self.ball_dx *= -1.05  # Slight speedup
                self.ball_dy += (self.ball_y - (self.p1_y + self.pad_h/2)) * 0.2
        elif self.ball_dx > 0 and self.ball_x >= self.w - self.pad_w - self.ball_size:
            if self.p2_y - self.ball_size <= self.ball_y <= self.p2_y + self.pad_h:
                self.ball_x = self.w - self.pad_w - self.ball_size
                self.ball_dx *= -1.05
                self.ball_dy += (self.ball_y - (self.p2_y + self.pad_h/2)) * 0.2
                
        # Speed cap
        self.ball_dx = max(-4.0, min(4.0, self.ball_dx))
        self.ball_dy = max(-3.0, min(3.0, self.ball_dy))
        
        # Scoring (Out of bounds)
        if self.ball_x < -10:
            self.reset_ball(left_served=True)
            self.force_miss = False
        elif self.ball_x > self.w + 10:
            self.reset_ball(left_served=False)
            self.force_miss = False
            
        # Draw paddles
        draw.rectangle([0, int(self.p1_y), self.pad_w - 1, int(self.p1_y) + self.pad_h], fill=(255, 255, 255))
        draw.rectangle([self.w - self.pad_w, int(self.p2_y), self.w - 1, int(self.p2_y) + self.pad_h], fill=(255, 255, 255))
        
        # Draw ball
        draw.rectangle([int(self.ball_x), int(self.ball_y), int(self.ball_x) + self.ball_size - 1, int(self.ball_y) + self.ball_size - 1], fill=(255, 255, 255))
        
        return img

File: src/engines/clocks/pong_clock.rs
use crate::core::matrix::MatrixBackend;
use crate::engines::renderers::BaseRenderer;
use rand::Rng;
use crate::engines::renderers::base_renderer::ArcadeFont;

pub struct PongClock {
    ball_x: f32,
    ball_y: f32,
    dx: f32,
    dy: f32,
    p1_y: f32,
    p2_y: f32,
    pad_h: f32,
    pad_w: i32,
    ball_size: i32,
    score_left: u32,  // Displays hours
    score_right: u32, // Displays minutes
    last_hour: i32,
    last_minute: i32,
    force_miss_left: bool,
    force_miss_right: bool,
}

impl PongClock {
    pub fn new(w: u32, h: u32) -> Self {
        let pad_h = (h as f32 / 3.0).max(8.0);
        let pad_w = ((w as f32 / 32.0) as i32).max(2);
        let ball_size = ((h as f32 / 16.0) as i32).max(2);
        let dx = (w as f32 / 40.0).max(1.5);
        let dy = (h as f32 / 30.0).max(1.0);
        Self {
            ball_x: w as f32 / 2.0,
            ball_y: h as f32 / 2.0,
            dx,
            dy,
            p1_y: h as f32 / 2.0,
            p2_y: h as f32 / 2.0,
            pad_h,
            pad_w,
            ball_size,
            score_left: 0,
            score_right: 0,
            last_hour: -1,
            last_minute: -1,
            force_miss_left: false,
            force_miss_right: false,
        }
    }

    fn reset_ball(&mut self, w: f32, h: f32, left_served: bool) {
        let mut rng = rand::thread_rng();
        self.ball_y = h / 2.0;
        self.dy = rng.gen_range(-1.5f32..1.5f32);
        if left_served {
            self.ball_x = self.pad_w as f32 + 2.0;
            self.dx = (w / 40.0).max(1.5);
        } else {
            self.ball_x = w - self.pad_w as f32 - 3.0;
            self.dx = -(w / 40.0).max(1.5);
        }
    }

    pub fn update_and_render(
        &mut self,
        matrix: &mut dyn MatrixBackend,
        hours: u32,
        minutes: u32,
        font: &ArcadeFont<'_>,
        scale: u32,
    ) {
        let w = matrix.width() as f32;
        let h = matrix.height() as f32;

        // Sync score with actual time
        if self.last_hour == -1 {
            self.last_hour = hours as i32;
            self.last_minute = minutes as i32;
            self.score_left = hours;
            self.score_right = minutes;
        }

        // Detect minute change → force a miss on the left paddle
        if minutes as i32 != self.last_minute {
            self.force_miss_left = true;
            self.last_minute = minutes as i32;
        }
        if hours as i32 != self.last_hour {
            self.last_hour = hours as i32;
        }

        // Physics update
        self.ball_x += self.dx;
        self.ball_y += self.dy;

        // Vertical wall bounce
        if self.ball_y <= 0.0 || self.ball_y >= h - 1.0 {
            self.dy = -self.dy;
            self.ball_y = self.ball_y.clamp(0.0, h - 1.0);
        }

        // AI: move paddles toward ball (with optional deliberate miss)
        let tracking_speed = h / 20.0;
        if self.force_miss_left {
            // Deliberate miss: move away from ball
            if self.ball_y < h / 2.0 {
                self.p1_y = (self.p1_y + tracking_speed).min(h - self.pad_h / 2.0);
            } else {
                self.p1_y = (self.p1_y - tracking_speed).max(self.pad_h / 2.0);
            }
        } else {
            let diff = self.ball_y - self.p1_y;
            self.p1_y += diff.signum() * tracking_speed.min(diff.abs());
        }

        let diff2 = self.ball_y - self.p2_y;
        self.p2_y += diff2.signum() * tracking_speed.min(diff2.abs());

        // Left paddle collision
        if self.ball_x <= self.pad_w as f32 + self.ball_size as f32 {
            let top = self.p1_y - self.pad_h / 2.0;
            let bot = self.p1_y + self.pad_h / 2.0;
            if self.force_miss_left || self.ball_y < top || self.ball_y > bot {
                // Miss! Point to right
                self.score_right = minutes;
                self.force_miss_left = false;
                self.reset_ball(w, h, false);
            } else {
                self.dx = self.dx.abs();
                self.ball_x = self.pad_w as f32 + self.ball_size as f32 + 1.0;
            }
        }

        // Right paddle collision
        if self.ball_x >= w - self.pad_w as f32 - self.ball_size as f32 {
            let top = self.p2_y - self.pad_h / 2.0;
            let bot = self.p2_y + self.pad_h / 2.0;
            if self.ball_y < top || self.ball_y > bot {
                // Miss! Point to left
                self.score_left = hours;
                self.reset_ball(w, h, true);
            } else {
                self.dx = -self.dx.abs();
                self.ball_x = w - self.pad_w as f32 - self.ball_size as f32 - 1.0;
            }
        }

        // ── Draw ──────────────────────────────────────────────────────────────

        // Dotted center line
        for y in (0..h as i32).step_by(2) {
            matrix.set_pixel((w / 2.0) as i32, y, 80, 80, 80);
        }

        // Left paddle
        let p1_top = (self.p1_y - self.pad_h / 2.0) as i32;
        let pad_thick = ((w / 128.0) as i32).max(2);
        for dy in 0..self.pad_h as i32 {
            for dx in 0..pad_thick {
                matrix.set_pixel(
                    self.pad_w + dx,
                    (p1_top + dy).clamp(0, h as i32 - 1),
                    255,
                    255,
                    255,
                );
            }
        }

        // Right paddle
        let p2_top = (self.p2_y - self.pad_h / 2.0) as i32;
        for dy in 0..self.pad_h as i32 {
            for dx in 0..pad_thick {
                matrix.set_pixel(
                    (w as i32) - self.pad_w - 1 - dx,
                    (p2_top + dy).clamp(0, h as i32 - 1),
                    255,
                    255,
                    255,
                );
            }
        }

        // Ball (square, ball_size × ball_size)
        for dy in 0..self.ball_size {
            for dx in 0..self.ball_size {
                matrix.set_pixel(
                    self.ball_x as i32 + dx,
                    self.ball_y as i32 + dy,
                    255,
                    255,
                    0,
                );
            }
        }

        // Scores drawn as TTF text at top center
        let score_left_str = format!("{:02}", self.score_left);
        let score_right_str = format!("{:02}", self.score_right);

        // Approximate width of 2 digits
        let digit_width = 15.0 * scale as f32;
        
        // Left score
        BaseRenderer::draw_text_at(
            matrix,
            &score_left_str,
            font,
            scale as f32,
            (w / 2.0 - digit_width - 4.0) as i32,
            4,
            (200, 200, 200),
            (0, 0, 0),
        );
        
        // Right score
        BaseRenderer::draw_text_at(
            matrix,
            &score_right_str,
            font,
            scale as f32,
            (w / 2.0 + 8.0) as i32,
            4,
            (200, 200, 200),
            (0, 0, 0),
        );
    }
}
